### PR TITLE
#17873: Listener socket rebinding to/from wildcard address when first bind() fails

### PR DESCRIPTION
**Pull request by rl1987**
https://trac.torproject.org/projects/tor/ticket/17873

### CODE REVIEW

File: src/core/mainloop/connection.c [hunk 1]
```diff
@@ -2759,20 +2844,54 @@ retry_all_listeners(smartlist_t *replaced_conns,
...
+
+    connection_t *new_conn =
+     connection_listener_new_for_port(r->new_port, &skip, &addr_in_use);
+    connection_t *old_conn = r->old_conn;
+
+    if (skip)
+      continue;
+
+    connection_close_immediate(old_conn);
+    connection_mark_for_close(old_conn);
```
Comment (ahf) [inline src/core/mainloop/connection.c]: Are both needed here?

File: src/core/mainloop/connection.c [hunk 2]
```diff
@@ -2759,20 +2844,54 @@ retry_all_listeners(smartlist_t *replaced_conns,
...
+#ifdef ENABLE_LISTENER_REBIND
+  SMARTLIST_FOREACH_BEGIN(replacements, listener_replacement_t *, r) {
+    int addr_in_use = 0;
+    int skip = 0;
+
+    tor_assert(r->new_port);
+    tor_assert(r->old_conn);
+
+    connection_t *new_conn =
+     connection_listener_new_for_port(r->new_port, &skip, &addr_in_use);
```
Comment (ahf) [inline src/core/mainloop/connection.c]: Nitpick: Add an additional space here.

File: src/core/mainloop/connection.c [hunk 3]
```diff
@@ -1473,6 +1506,8 @@ connection_listener_new(const struct sockaddr *listensockaddr,
    */
   connection_check_oos(get_n_open_sockets(), 0);
 
+  log_notice(LD_NET, "Opened %s on %s",
+             conn_type_to_string(type), fmt_addrport(&addr, usePort));
```
Comment (ahf) [inline src/core/mainloop/connection.c]: Should this maybe be at info log level instead?

File: src/core/mainloop/connection.c (hunk 2 above)
Comment (rl1987) [reply src/core/mainloop/connection.c]: 1d92370f40475546f7b672a444cfc777b3cb577c

File: src/core/mainloop/connection.c (hunk 1 above)
Comment (rl1987) [reply src/core/mainloop/connection.c]: I believe so - we want to close the old connection right away, without trying to send any data from its write buffer. It is generally done like this in connection.c when closing no longer configured or failed listeners.

See:
* https://gitweb.torproject.org/tor.git/tree/src/core/mainloop/connection.c#n2773
* https://gitweb.torproject.org/tor.git/tree/src/core/mainloop/connection.c#n3426

File: src/core/mainloop/connection.c (hunk 3 above)
Comment (nmathewson) [reply src/core/mainloop/connection.c]: I think we have it at notice now, right?  We can leave it there if so, and move it to info later if we decide.

(If we don't have this at notice now, please ignore me)

**Pull request closed without merging**